62.SchusslerTest/shader.frag

```diff
@@ -0,0 +1,127 @@
+#version 430 core
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+
+layout (location = 0) in vec3 Normal;
+layout (location = 1) in vec3 Pos;
+layout (location = 2) flat in vec3 LightPos;
+layout (location = 3) flat in vec3 CamPos;
+layout (location = 4) flat in vec3 GNormal;
+layout (location = 5) flat in float Alpha;
+layout (location = 6) flat in float AlphaY;
+
+layout (location = 0) out vec4 outColor;
+
+layout (push_constant) uniform PC {
+    layout (offset = 64) vec3 campos;
+    layout (offset = 80) uint testNum;
+} pc;
+
+#define TEST_GGX 1
+#define TEST_BECKMANN 2
+#define TEST_PHONG 3
+#define TEST_AS 4
+#define TEST_OREN_NAYAR 5
+#define TEST_LAMBERT 6
+
+#include <nbl/builtin/glsl/bxdf/brdf/specular/ggx.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/specular/beckmann.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/specular/blinn_phong.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/diffuse/oren_nayar.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/diffuse/lambert.glsl>
+
+float pdot(vec3 a, vec3 b) {
+    return max(0.0, dot(a,b));
+}
+
+float ap(vec3 wi, vec3 wp, vec3 wg) {
+    return pdot(wi, wp)/pdot(wp, wg);
+}
+
+float at(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return pdot(wi, wt) * sqrt(1-(pdot(wp,wg)*pdot(wp,wg))) / pdot(wp,wg);
+}
+
+float lambdap(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return ap(wi, wp, wg) / (ap(wi,wp,wg) + at(wi,wt,wp,wg));
+}
+
+float lambdat(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return at(wi,wt,wp,wg) / (ap(wi,wp,wg) + at(wi,wt,wp,wg));
+}
```

Comment:
`at` and `ap` both have a denominator `pdot(wp,wg)` which actually happens to be the `deviation`

since its in all of them there's no point applying it.

Reply:
furthermore `(ap+at)*deviation` can be optimized to
```
pdot(wi,wp)+pdot(wi,wt)*sqrt(1.f-deviation*deviation)
```

but since
```
wt = normalize(wg*deviation-wp) = (wg*deviation-wp)/length(wg*deviation-wp) = (wg*deviation-wp)/sqrt(1.f-deviation*deviation)
```
we have
```
pdot(wi,wp)+pdot(wi,wg*deviation-wp) == pdot(wi,wp)+pdot(wi,unnormalized_wt)
```

Reply:
this way everything is much more numerically robust because it will work even when `deviation=0.f` since we don't have divisions by 0 anymore (unless `wi` is below the geometrical hemisphere, which is when we should never even start to shade the surface in the first place)

Reply:
Also to make your life even more simpler `lambda_t = 1.f-lambda_p` :D

Reply:
In https://github.com/Devsh-Graphics-Programming/Nabla/issues/310 I call `lambda_p`, `shadingNormal_vpdf` instead.

The two definitions/derivations are compatible. This is the most optimal
```glsl
float nbl_glsl_bxdf_Schussler_shadingNormalVPDF(in float NdotX, in float clampedNdotX, in float GdotX, in float GdotN/*old deviation*/)
{
   assert(GeomNdotShadingN>=0.f);
   assert(XdotGeomN>0.f);
   return clampedNdotX/(clampedNdotX+max(NdotX-GdotN*GdotX,0.f));
}
```

We could define a
```glsl
// a_p/(a_p+a_t)
float nbl_glsl_bxdf_schussler_shadingNormal_vpdf(in float clampedNdotX, in float DenormTdotX)
{
   return NdotX/(NdotX+DenormTdotX);
}
```

Reply:
Bonus points: for BSDFs we'd need to invert the sign of `NdotX` and `GdotX` whenever `GdotX<0.f`

Assuming that for BRDFs we check and skip all shading logic when `GdotX`, this means we can `abs` it in the above function